plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergRegisterTableProcedure.java

```diff
@@ -0,0 +1,325 @@
+        metastore = createTestingFileHiveMetastore(metastoreDir);
+        trackingFileSystemFactory = new TrackingFileSystemFactory(new HdfsFileSystemFactory(HDFS_ENVIRONMENT));
+        trinoFileSystem = trackingFileSystemFactory.create(TestingConnectorSession.SESSION);
+        return IcebergQueryRunner.builder()
+                .setMetastoreDirectory(metastoreDir)
+                .build();
+    }
+
+    @BeforeClass
+    @Override
+    public void init()
+            throws Exception
+    {
+        super.init();
+        assertUpdate(format("CREATE SCHEMA %s", TEST_SCHEMA_NAME));
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void tearDown()
+            throws IOException
+    {
+        deleteRecursively(metastoreDir.toPath(), ALLOW_INSECURE);
+    }
+
+    @DataProvider
+    public static Object[][] storageFormats()
+    {
+        return Stream.of(StorageFormat.values())
+                .map(storageFormat -> new Object[] {storageFormat})
+                .toArray(Object[][]::new);
+    }
+
+    @Test(dataProvider = "storageFormats")
+    public void testRegisterTableWithTableLocation(StorageFormat storageFormat)
+    {
+        String baseTableName = "test_register_table_with_table_location_" + storageFormat + "_" + randomTableSuffix();
+        String trinoTableName = getTrinoTableName(baseTableName);
+
+        assertUpdate(format("CREATE TABLE %s (a int, b varchar, c boolean) with (format = '%s')", trinoTableName, storageFormat));
+        assertUpdate(format("INSERT INTO %s values(1, 'INDIA', true)", trinoTableName), 1);
+        assertUpdate(format("INSERT INTO %s values(2, 'USA', false)", trinoTableName), 1);
+        assertThat(query(format("SELECT * FROM %s", trinoTableName)))
+                .matches("VALUES ROW(INT '1', VARCHAR 'INDIA', BOOLEAN 'true'), ROW(INT '2', VARCHAR 'USA', BOOLEAN 'false')");
+
+        String tableLocation = getTableLocation(trinoTableName);
+        String showCreateTableOld = (String) computeActual("SHOW CREATE TABLE " + trinoTableName).getOnlyValue();
+
+        // Drop table from hive metastore and use the same table name to register again with the metadata
+        metastore.dropTable(TEST_SCHEMA_NAME, baseTableName, false);
+        assertThat(metastore.getTable(TEST_SCHEMA_NAME, baseTableName)).as("Table in metastore should be dropped").isEmpty();
+
+        assertUpdate(format("CALL iceberg.system.register_table ('%s', '%s', '%s')", TEST_SCHEMA_NAME, baseTableName, tableLocation));
+        String showCreateTableNew = (String) computeActual("SHOW CREATE TABLE " + trinoTableName).getOnlyValue();
+
+        assertThat(query(format("SELECT * FROM %s", trinoTableName)))
+                .matches("VALUES ROW(INT '1', VARCHAR 'INDIA', BOOLEAN 'true'), ROW(INT '2', VARCHAR 'USA', BOOLEAN 'false')");
+        assertThat(showCreateTableOld).isEqualTo(showCreateTableNew);
+        assertUpdate(format("DROP TABLE %s", trinoTableName));
```

Comment:
Let's add a DML operation after registering the table just to make sure that the newly registered table can still be modified.